File: bot.py
import requests
import json
import time
import base64
from concurrent.futures import ThreadPoolExecutor
from rich.console import Console
from rich.logging import RichHandler
from rich.progress import Progress, TextColumn, BarColumn, TimeRemainingColumn
import logging
from typing import Tuple, Optional, List, Dict

# Inisialisasi console untuk output rich
console = Console()

# Konfigurasi logging dengan RichHandler
logging.basicConfig(
    level=logging.INFO,
    format="%(message)s",
    datefmt="[%Y-%m-%d %H:%M:%S]",
    handlers=[RichHandler(rich_tracebacks=True, console=console)]
)
logger = logging.getLogger("TaskBot")

BASE_HEADERS = {
    "accept": "application/json, text/plain, */*",
    "accept-encoding": "gzip, deflate, br, zstd",
    "accept-language": "en-US,en;q=0.9",
    "cache-control": "no-cache",
    "content-type": "application/json",
    "origin": "https://tg-wallet.planx.io",
    "pragma": "no-cache",
    "priority": "u=1, i",
    "referer": "https://tg-wallet.planx.io/",
    "sec-ch-ua": '"Chromium";v="133", "Microsoft Edge WebView2";v="133", "Not(A:Brand";v="99", "Microsoft Edge";v="133"',
    "sec-ch-ua-mobile": "?0",
    "sec-ch-ua-platform": '"Windows"',
    "sec-fetch-dest": "empty",
    "sec-fetch-mode": "cors",
    "sec-fetch-site": "same-site",
    "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/133.0.0.0 Safari/537.36 Edg/133.0.0.0"
}

TASKS = {
    "m20250212173935584402900025": "Daily Login",
    "m20250212173935594680500028": "Follow Twitter",
    "m20250212173935519374200019": "Join Telegram",
    "m20250212173935604389100031": "Share Post",
    "m20250212173935613755700034": "Invite Friend",
    "m20250212173935571986800022": "Watch Video",
    "m20250214173952165258600005": "Complete Survey",
    "m20250213173941632390600015": "Visit Website",
    "m20250213173941720460300018": "Like Post",
    "m20250214173952169399300006": "Comment Post",
    "m20250213173941728955700021": "Daily Check-in",
    "m20250213173941736560000024": "Play Game",
    "m20250213173941767785900027": "Submit Feedback",
    "m20250212173934013124700001": "Profile Setup"
}

CALL_URL = "https://mpc-api.planx.io/api/v1/telegram/task/call"
CLAIM_URL = "https://mpc-api.planx.io/api/v1/telegram/task/claim"

def decode_jwt_token(token: str) -> Tuple[Optional[str], Optional[str]]:
    """
    Decode JWT token untuk mendapatkan user_id dari payload.

    Args:
        token (str): Token JWT yang akan didecode.

    Returns:
        Tuple[Optional[str], Optional[str]]: (user_id, token) atau (None, None) jika gagal.
    """
    try:
        parts = token.split('.')
        if len(parts) != 3:
            logger.error("Token JWT tidak valid, harus terdiri dari 3 bagian")
            return None, None
        
        payload = parts[1]
        decoded_payload = base64.urlsafe_b64decode(payload + '=' * (-len(payload) % 4)).decode('utf-8')
        payload_data = json.loads(decoded_payload)
        
        user_id = payload_data.get("user_id")
        if not user_id:
            logger.error("user_id tidak ditemukan di payload token")
            return None, None
        
        return user_id, token
    except Exception as e:
        logger.error(f"Error decoding token: {e}", exc_info=True)
        return None, None

def call_task(user_id: str, token: str, task_id: str, task_name: str) -> bool:
    """
    Melakukan panggilan tugas.

    Args:
        user_id (str): ID pengguna.
        token (str): Token autentikasi.
        task_id (str): ID tugas.
        task_name (str): Nama tugas.

    Returns:
        bool: True jika tugas berhasil, False jika gagal.
    """
    headers = BASE_HEADERS.copy()
    headers["token"] = token
    payload = {"taskId": task_id}
    
    logger.info(f"[{user_id}] Mengerjakan tugas: {task_name}")
    try:
        response = requests.post(CALL_URL, headers=headers, json=payload, timeout=10)
        data = response.json()
        
        if response.status_code == 200 and data.get("success", False):
            logger.info(f"[{user_id}] Tugas {task_name} berhasil dikerjakan")
            return True
        else:
            logger.warning(f"[{user_id}] Gagal mengerjakan tugas {task_name} - Code: {data.get('code', response.status_code)}, Msg: {data.get('msg', 'Unknown error')}")
            return False
    except requests.RequestException as e:
        logger.error(f"[{user_id}] Error saat mengerjakan tugas {task_name}: {e}", exc_info=True)
        return False

def claim_task(user_id: str, token: str, task_id: str, task_name: str) -> bool:
    """
    Mengklaim hadiah tugas.

    Args:
        user_id (str): ID pengguna.
        token (str): Token autentikasi.
        task_id (str): ID tugas.
        task_name (str): Nama tugas.

    Returns:
        bool: True jika klaim berhasil, False jika gagal.
    """
    headers = BASE_HEADERS.copy()
    headers["token"] = token
    payload = {"taskId": task_id}
    
    logger.info(f"[{user_id}] Mengklaim hadiah untuk tugas: {task_name}")
    try:
        response = requests.post(CLAIM_URL, headers=headers, json=payload, timeout=10)
        data = response.json()
        
        if response.status_code == 200 and data.get("success", False):
            logger.info(f"[{user_id}] Hadiah untuk tugas {task_name} berhasil diklaim")
            return True
        else:
            code = data.get('code', response.status_code)
            msg = data.get('msg', 'Unknown error')
            if code == 1015:
                logger.warning(f"[{user_id}] Tugas {task_name} sudah pernah diklaim sebelumnya")
            else:
                logger.warning(f"[{user_id}] Gagal mengklaim hadiah {task_name} - Code: {code}, Msg: {msg}")
            return False
    except requests.RequestException as e:
        logger.error(f"[{user_id}] Error saat mengklaim hadiah {task_name}: {e}", exc_info=True)
        return False

def process_account(token: str) -> None:
    """
    Proses semua tugas untuk satu akun: call dulu semua, lalu claim.

    Args:
        token (str): Token autentikasi untuk akun.
    """
    try:
        token = token.strip()
        if not token or not token.startswith("Bearer "):
            logger.error(f"Token tidak valid: {token}")
            return
        
        user_id, token = decode_jwt_token(token)
        if not user_id or not token:
            logger.error(f"Gagal memproses token: {token}")
            return
            
        logger.info(f"\nMemproses akun dengan ID: {user_id}")
        
        # Tahap 1: Call semua tugas
        successful_tasks: List[Tuple[str, str]] = []
        for task_id, task_name in TASKS.items():
            if call_task(user_id, token, task_id, task_name):
                successful_tasks.append((task_id, task_name))
            time.sleep(2)  # Delay antar tugas
        
        # Tahap 2: Claim semua tugas yang berhasil
        if successful_tasks:
            logger.info(f"\n[{user_id}] Mulai mengklaim {len(successful_tasks)} tugas yang berhasil")
            for task_id, task_name in successful_tasks:
                claim_task(user_id, token, task_id, task_name)
                time.sleep(2)  # Delay antar klaim
        else:
            logger.warning(f"[{user_id}] Tidak ada tugas yang berhasil untuk diklaim")
        
        logger.info(f"Selesai memproses akun {user_id}\n")
    
    except Exception as e:
        logger.error(f"Error memproses akun: {e}", exc_info=True)

def wait_with_progress_bar(wait_time_seconds: int) -> None:
    """
    Menampilkan progress bar selama waktu tunggu.

    Args:
        wait_time_seconds (int): Waktu tunggu dalam detik.
    """
    with Progress(
        TextColumn("[progress.description]{task.description}"),
        BarColumn(),
        TimeRemainingColumn(),
        console=console
    ) as progress:
        task = progress.add_task("[cyan]Menunggu siklus berikutnya...", total=wait_time_seconds)
        elapsed = 0
        while elapsed < wait_time_seconds:
            time.sleep(1)  # Update setiap detik
            elapsed += 1
            progress.update(task, advance=1)

def main_loop(max_threads: int) -> None:
    """
    Main loop yang berjalan setiap 3 jam 10 menit.

    Args:
        max_threads (int): Jumlah maksimum thread untuk ThreadPoolExecutor.
    """
    while True:
        try:
            with open("data.txt", "r") as file:
                tokens = file.readlines()
            
            if not tokens:
                logger.warning("Tidak ada token di data.txt!")
                wait_with_progress_bar(190 * 60)  # 3 jam 10 menit
                continue
                
            logger.info(f"\nMulai proses untuk {len(tokens)} akun pada {time.ctime()} dengan {max_threads} threads")
            
            # Proses semua akun secara paralel dengan jumlah thread dari input
            with ThreadPoolExecutor(max_workers=max_threads) as executor:
                futures = [executor.submit(process_account, token) for token in tokens]
                for future in futures:
                    try:
                        future.result()  # Menunggu hasil dan menangani exception
                    except Exception as e:
                        logger.error(f"Error dalam thread: {e}", exc_info=True)
                
            logger.info(f"\nSelesai satu siklus.")
            wait_with_progress_bar(190 * 60)  # 3 jam 10 menit = 190 menit
            
        except FileNotFoundError:
            logger.error("File data.txt tidak ditemukan! Melanjutkan loop setelah delay...")
            wait_with_progress_bar(190 * 60)
        except KeyboardInterrupt:
            logger.warning("Program dihentikan oleh pengguna (KeyboardInterrupt), melanjutkan loop...")
            wait_with_progress_bar(190 * 60)
        except Exception as e:
            logger.error(f"Error di main loop: {e}", exc_info=True)
            wait_with_progress_bar(190 * 60)

if __name__ == "__main__":
    # Tampilkan header tabel dengan rich
    console.print("[bold cyan]╔══════════════════════════════════════════════╗[/bold cyan]")
    console.print("[bold cyan]║       🌟 PlanX BOT - Automated Rewards       ║[/bold cyan]")
    console.print("[bold cyan]║   Automate your PlanX task completion!       ║[/bold cyan]")
    console.print("[bold cyan]║  Developed by: https://t.me/sentineldiscus   ║[/bold cyan]")
    console.print("[bold cyan]╚══════════════════════════════════════════════╝[/bold cyan]")
    console.print("[bold green]Memulai program...[/bold green]")

    while True:
        try:
            max_threads = int(console.input("[bold yellow]Masukkan jumlah threads yang diinginkan: [/bold yellow]"))
            if max_threads <= 0:
                console.print("[bold red]Jumlah threads harus lebih dari 0![/bold red]")
                continue
            break
        except ValueError:
            console.print("[bold red]Masukkan angka yang valid![/bold red]")
    
    console.print(f"[bold green]Program akan berjalan dengan {max_threads} threads[/bold green]")
    try:
        main_loop(max_threads)
    except KeyboardInterrupt:
        console.print("[bold red]Program dihentikan oleh pengguna melalui KeyboardInterrupt.[/bold red]")
    except Exception as e:
        logger.error(f"Error kritis di luar loop: {e}", exc_info=True)

File: bot.rb
require 'httparty'
require 'colorize'
require 'securerandom'
require 'json'
require 'uri'
require 'jwt'
require 'digest'
require 'fileutils'

puts <<~BANNER
╔══════════════════════════════════════════════╗
║       🌟 PlanX TaskBot - Automated Tasks     ║
║   Automate your PlanX account tasks!         ║
║  Developed by: https://t.me/sentineldiscus   ║
╚══════════════════════════════════════════════╝
BANNER

TASKS = {
  "m20250212173934013124700001" => "Daily Login",
  "m20250325174288367185100003" => "Lottery",
  "m20250522174789949352000045" => "I am not a robot",
  "m20250521174781856608400042" => "Earn 1.7% Daily with USDX in Xwallet!",
  "m20250521174780862795300039" => "Earn TON for completing simple tasks",
  "m20250521174779954000200036" => "Join RewardsHQ and win cash prizes",
  "m20250521174779685918100032" => "Search BTC wallets with RootBTC",
  "m20250521174779639447100029" => "Join TAPX & Earn Rewards",
  "m20250521174779307171300025" => "Play & Earn Your $1000",
  "m20250520174772991278700024" => "Play Merge Pals and Earn Rewards",
  "m20250520174770655580300018" => "Play Valor Quest and Earn Rewards",
  "m20250520174770687133400021" => "Explore Funton.ai TG Mini App and Earn Rewards",
  "m20250520174770621353700015" => "Post Short Videos and Earn Like TikTok",
  "m20250516174736524661500009" => "Mine for free and win $300",
  "m20250515174730236703400006" => "Start Rolling - Play DiceSwap & Win TON",
  "m20250515174729364086600003" => "Launch TapCoinsBot",
  "m20250513174712786995700006" => "Play Miner to win $300",
  "m20250513174712489462800003" => "Let's earn $WTON and WONTON collections together",
  "m20250424174548205765500003" => "Earn USDT Daily",
  "m20250212173935146770000006" => "Create Wallet",
  "m20250507174660749235200006" => "Visit the PlanX Official Website",
  "m20250507174659651577000003" => "Play Startai, Claim 100 USDT",
  "m20250505174642807898300003" => "Play Simple and get $SMPL",
  "m20250212173935519374200019" => "Join the PlanX Community",
  "m20250212173935571986800022" => "Join the PlanX Channel",
  "m20250212173935594680500028" => "Follow PlanX on X",
  "m20250212173935584402900025" => "Join the PlanX Discord",
  "m20250212173935604389100031" => "Follow PlanX on TikTok",
  "m20250212173935613755700034" => "Follow PlanX on YouTube",
  "m20250214173952165258600005" => "Repost a PlanX'post on X",
  "m20250213173941632390600015" => "Comment a PlanX'post on X",
  "m20250213173941720460300018" => "Like a PlanX'post on X",
  "m20250214173952169399300006" => "Quote a PlanX' post and tag 3 of friends on X",
  "m20250213173941728955700021" => "Share the PlanX video from YouTube to X",
  "m20250213173941736560000024" => "Share the PlanX video from TikTok to X",
  "m20250213173941767785900027" => "Read the PlanX Medium article"
}

CLAIM_ONLY_TASKS = {
  "m20250212173934013124700001" => "Daily Login",
  "m20250325174288367185100003" => "Lottery"
}

CALL_TASKS = TASKS.reject { |k, _| CLAIM_ONLY_TASKS.key?(k) }

DEFAULT_USER_AGENTS_AND_PLATFORMS = [
  {
    user_agent: "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36",
    platform: "Windows"
  },
  {
    user_agent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36",
    platform: "macOS"
  },
  {
    user_agent: "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36",
    platform: "Linux"
  },
  {
    user_agent: "Mozilla/5.0 (Linux; Android 14; SM-G998B) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Mobile Safari/537.36",
    platform: "Android"
  },
  {
    user_agent: "Mozilla/5.0 (iPhone; CPU iPhone OS 17_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Mobile/15E148 Safari/604.1",
    platform: "iOS"
  }
]

HEADERS = {
  'accept' => 'application/json, text/plain, */*',
  'accept-encoding' => 'identity',
  'content-type' => 'application/json',
  'language' => 'id',
  'origin' => 'https://tg-wallet.planx.io',
  'referer' => 'https://tg-wallet.planx.io/'
}

USE_PROXY = false

def generate_fp(token, user_agent)
  timestamp = Time.now.to_i.to_s
  screen_resolution = "1920x1080"
  timezone = "Asia/Jakarta"
  input = "#{token}:#{user_agent}:#{screen_resolution}:#{timezone}:#{timestamp}"
  Digest::MD5.hexdigest(input)
end

def read_or_generate_fps(tokens, user_agents)
  FileUtils.mkdir_p('HEADERS')
  fp_file = 'HEADERS/fp.json'
  fps = []
  if File.exist?(fp_file)
    begin
      fps = JSON.parse(File.read(fp_file))
    rescue StandardError => e
      puts "Error reading #{fp_file}: #{e.message}".red
    end
  end
  tokens.each_with_index do |token, i|
    fps[i] ||= generate_fp(token, user_agents[i])
  end
  begin
    File.write(fp_file, JSON.pretty_generate(fps))
  rescue StandardError => e
    puts "Error saving #{fp_file}: #{e.message}".red
  end
  fps
end

def read_or_generate_user_agents_and_platforms(token_count)
  FileUtils.mkdir_p('HEADERS')
  ua_file = 'HEADERS/useragent.json'
  platform_file = 'HEADERS/platform.json'
  user_agents = []
  platforms = []
  if File.exist?(ua_file) && File.exist?(platform_file)
    begin
      user_agents = JSON.parse(File.read(ua_file))
      platforms = JSON.parse(File.read(platform_file))
    rescue StandardError => e
      puts "Error reading #{ua_file} or #{platform_file}: #{e.message}".red
    end
  end
  while user_agents.size < token_count
    pair = DEFAULT_USER_AGENTS_AND_PLATFORMS[user_agents.size % DEFAULT_USER_AGENTS_AND_PLATFORMS.size]
    user_agents << pair[:user_agent]
    platforms << pair[:platform]
  end
  begin
    File.write(ua_file, JSON.pretty_generate(user_agents))
    File.write(platform_file, JSON.pretty_generate(platforms))
  rescue StandardError => e
    puts "Error saving #{ua_file} or #{platform_file}: #{e.message}".red
  end
  [user_agents, platforms]
end

def decode_token(token)
  begin
    decoded = JWT.decode(token, nil, false)
    decoded[0]['username'] || 'Unknown'
  rescue StandardError => e
    puts "Error decoding token: #{e.message}".red
    'Unknown'
  end
end

def read_tokens
  begin
    tokens = File.exist?('data.txt') ? File.readlines('data.txt').map(&:strip).reject(&:empty?) : []
    tokens.map do |token|
      token = token[7..-1].strip if token.downcase.start_with?('bearer ')
      if token.empty?
        puts "Empty token detected".red
        nil
      else
        token
      end
    end.compact
  rescue StandardError => e
    puts "Error: data.txt not found or unreadable: #{e.message}".red
    []
  end
end

def read_proxies
  return [] unless USE_PROXY && File.exist?('proxy.txt')
  begin
    File.readlines('proxy.txt').map(&:strip).reject(&:empty?)
  rescue StandardError => e
    puts "Error reading proxy.txt: #{e.message}".red
    []
  end
end

def format_proxy(proxy)
  return nil if proxy.nil? || proxy.empty?
  proxy = "http://#{proxy}" unless proxy.include?('://')
  begin
    uri = URI.parse(proxy)
    unless uri.scheme && uri.host && uri.port
      puts "Invalid proxy format: #{proxy}".red
      return nil
    end
    proxy_options = {
      http_proxyaddr: uri.host,
      http_proxyport: uri.port
    }
    proxy_options[:http_proxyuser] = uri.user if uri.user
    proxy_options[:http_proxypass] = uri.password if uri.password
    proxy_options
  rescue URI::InvalidURIError => e
    puts "Error parsing proxy #{proxy}: #{e.message}".red
    nil
  end
end

def check_task_status(token, fp, user_agent, platform, proxy = nil)
  url = 'https://mpc-api.planx.io/api/v1/telegram/task/list'
  headers = HEADERS.merge(
    'token' => "Bearer #{token}",
    'fp' => fp,
    'user-agent' => user_agent
  )
  options = { headers: headers }
  options.merge!(proxy) if proxy
  begin
    response = HTTParty.get(url, options.merge(timeout: 10))
    parsed_response = JSON.parse(response.body) rescue nil
    if response.code == 200 && parsed_response && parsed_response['success']
      parsed_response['data'] || []
    else
      puts "Failed to fetch task status".red
      []
    end
  rescue StandardError => e
    puts "Failed to fetch task status: #{e.message}".red
    []
  end
end

def call_task(task_id, token, fp, user_agent, platform, proxy = nil)
  url = 'https://mpc-api.planx.io/api/v1/telegram/task/call'
  headers = HEADERS.merge(
    'token' => "Bearer #{token}",
    'fp' => fp,
    'user-agent' => user_agent
  )
  payload = { taskId: task_id }
  options = { headers: headers, body: payload.to_json }
  options.merge!(proxy) if proxy
  begin
    response = HTTParty.post(url, options.merge(timeout: 10))
    parsed_response = JSON.parse(response.body) rescue nil
    if response.code == 200 && parsed_response && parsed_response['success']
      puts "Task #{TASKS[task_id]} succeeded".green
      true
    elsif response.code == 200 && parsed_response && parsed_response['msg'] == 'Frequent access, please try again later'
      puts "Task #{TASKS[task_id]} rate limited. Waiting 15 seconds...".yellow
      sleep 15
      false
    elsif response.code == 406 || (parsed_response && parsed_response['code'] == 406)
      puts "Task #{TASKS[task_id]} already completed".blue
      true
    else
      error_message = parsed_response&.[]('msg')
      if error_message && !error_message.empty?
        puts "Task #{TASKS[task_id]} failed: #{error_message}".red
      else
        puts "Task #{TASKS[task_id]} failed".red
      end
      false
    end
  rescue StandardError => e
    puts "Task #{TASKS[task_id]} failed: #{e.message}".red
    false
  end
end

def claim_task(task_id, token, fp, user_agent, platform, proxy = nil, task_list = TASKS)
  url = 'https://mpc-api.planx.io/api/v1/telegram/task/claim'
  headers = HEADERS.merge(
    'token' => "Bearer #{token}",
    'fp' => fp,
    'user-agent' => user_agent
  )
  payload = { taskId: task_id }
  options = { headers: headers, body: payload.to_json }
  options.merge!(proxy) if proxy
  begin
    response = HTTParty.post(url, options.merge(timeout: 10))
    parsed_response = JSON.parse(response.body) rescue nil
    if response.code == 200 && parsed_response && parsed_response['success']
      puts "Claim task #{task_list[task_id]} succeeded".green
      true
    elsif response.code == 200 && parsed_response && parsed_response['msg'] == 'Frequent access, please try again later'
      puts "Claim task #{task_list[task_id]} rate limited. Waiting 15 seconds...".yellow
      sleep 15
      false
    elsif response.code == 406 || (parsed_response && parsed_response['code'] == 406)
      puts "Task #{task_list[task_id]} already claimed".blue
      true
    else
      error_message = parsed_response&.[]('msg')
      if error_message && !error_message.empty?
        puts "Claim task #{task_list[task_id]} failed: #{error_message}".red
      else
        puts "Claim task #{task_list[task_id]} failed".red
      end
      false
    end
  rescue StandardError => e
    puts "Claim task #{task_list[task_id]} failed: #{e.message}".red
    false
  end
end

def main
  puts "Starting PlanX TaskBot...".yellow
  iteration = 1
  loop do
    tokens = read_tokens
    if tokens.empty?
      puts "No valid tokens found. Script stopped.".red
      return
    end
    user_agents, platforms = read_or_generate_user_agents_and_platforms(tokens.size)
    if user_agents.size < tokens.size || platforms.size < tokens.size
      puts "Insufficient user-agents or platforms for all tokens. Script stopped.".red
      return
    end
    fps = read_or_generate_fps(tokens, user_agents)
    if fps.size < tokens.size
      puts "Insufficient FPs for all tokens. Script stopped.".red
      return
    end
    proxies = read_proxies
    tokens.each_with_index do |token, i|
      username = decode_token(token)
      puts "\nAccount #{username}: Starting...".yellow
      fp = fps[i]
      user_agent = user_agents[i]
      platform = platforms[i]
      proxy = proxies.any? ? format_proxy(proxies.sample) : nil
      puts "Account #{username}: Using FP #{fp}".blue
      puts "Account #{username}: Using User-Agent #{user_agent}".blue
      puts "Account #{username}: Using Platform #{platform}".blue
      puts "Account #{username}: Using proxy #{proxy[:http_proxyaddr]}:#{proxy[:http_proxyport]}" if proxy
      puts "Account #{username}: Checking task status...".yellow
      task_statuses = check_task_status(token, fp, user_agent, platform, proxy)
      if iteration == 1
        puts "Account #{username}: Processing CALL tasks...".yellow
        CALL_TASKS.each do |task_id, _|
          task_status = task_statuses.find { |t| t['taskId'] == task_id }
          if task_status && task_status['completed']
            puts "Task #{TASKS[task_id]} already completed, skipping call".blue
            next
          end
          call_task(task_id, token, fp, user_agent, platform, proxy)
          sleep 0
        end
        puts "Account #{username}: Waiting 30 seconds before CLAIM...".yellow
        sleep 30
        puts "Account #{username}: Processing CLAIM tasks...".yellow
        TASKS.each do |task_id, _|
          task_status = task_statuses.find { |t| t['taskId'] == task_id }
          if task_status && task_status['claimed']
            puts "Task #{TASKS[task_id]} already claimed, skipping claim".blue
            next
          end
          claim_task(task_id, token, fp, user_agent, platform, proxy)
          sleep 0
        end
      else
        puts "Account #{username}: Processing CLAIM tasks (Daily Login & Lottery)...".yellow
        CLAIM_ONLY_TASKS.each do |task_id, _|
          task_status = task_statuses.find { |t| t['taskId'] == task_id }
          if task_status && task_status['claimed']
            puts "Task #{CLAIM_ONLY_TASKS[task_id]} already claimed, skipping claim".blue
            next
          end
          claim_task(task_id, token, fp, user_agent, platform, proxy, CLAIM_ONLY_TASKS)
          sleep 5
        end
      end
      if i < tokens.size - 1
        puts "Account #{username}: Done. Waiting 3 seconds before next account...".yellow
        sleep 3
      end
    end
    iteration += 1
    puts "All accounts processed. Waiting 3 hours before next iteration...".yellow
    sleep 10800
  end
end

main if __FILE__ == $PROGRAM_NAME
